src/contract.rs

```diff
@@ -0,0 +1,412 @@
+            proposer: msg::source(),
+            applicant: *applicant,
+            quorum,
+            amount,
+            details,
+            starting_period,
+            ended_at: starting_period + self.voting_period_length,
+            ..Proposal::default()
+        };
+
+        self.proposals.insert(self.proposal_id, proposal);
+
+        msg::reply(
+            DaoEvent::SubmitFundingProposal {
+                proposer: msg::source(),
+                applicant: *applicant,
+                proposal_id: self.proposal_id,
+                amount,
+            },
+            0,
+        )
+        .unwrap();
+        self.proposal_id = self.proposal_id.saturating_add(1);
+        self.locked_funds = self.locked_funds.saturating_add(amount);
+    }
+
+    /// The member submit his vote (YES or NO) on the proposal
+    /// Requirements:
+    /// * The proposal can be submitted only by the existing members or their delegate addresses
+    /// * The member can vote on the proposal only once
+    /// * Proposal must exist, the voting period must has started and not expired
+    /// Arguments:
+    /// * `proposal_id`: the proposal ID
+    /// * `vote`: the member  a member vote (YES or NO)
+    fn submit_vote(&mut self, proposal_id: u128, vote: Vote) {
+        self.check_for_membership();
+
+        // checks that proposal exists, the voting period has started, not expired and that member did not vote on the proposal
+        let proposal = match self.proposals.get_mut(&proposal_id) {
+            Some(proposal) => {
+                if exec::block_timestamp() > proposal.starting_period + self.voting_period_length {
+                    panic!("proposal voting period has expired");
+                }
+                if exec::block_timestamp() < proposal.starting_period {
+                    panic!("voting period has not started");
+                }
+                if proposal
+                    .votes_by_member
+                    .iter()
+                    .any(|(actor, _vote)| msg::source().eq(actor))
+                {
+                    panic!("account has already voted on that proposal");
+                }
+                proposal
+            }
+            None => {
+                panic!("proposal does not exist");
+            }
+        };
+
+        let member = self.members.get_mut(&msg::source()).unwrap();
+
+        match vote {
+            Vote::Yes => {
+                proposal.yes_votes = proposal.yes_votes.saturating_add(member.shares);
+                // it is necessary to save the highest id of the proposal - must be processed for member to ragequit
+                if let Some(id) = member.highest_index_yes_vote {
+                    if id < proposal_id {
+                        member.highest_index_yes_vote = Some(proposal_id);
+                    }
+                } else {
+                    member.highest_index_yes_vote = Some(proposal_id);
+                }
```

Comment:
Might be shorted and clearer:
```suggestion
                let id = member.highest_index_yes_vote.get_or_insert(proposal_id);
                *id = proposal_id.max(*id);
```